File: Numbers/Pi/Python/pi.py
# Найти число пи до n-й цифры после запятой
# Find Pi up to the Nth digit
import math

MAX_NUMBER = 10000


def sqrt(n, limit):
    """
    n, limit -> x
    Calculate square of number (n) with limit
    """
    precision = 10 ** 16
    n_float = float((n * precision) // limit) / precision
    x = (int(precision * math.sqrt(n_float)) * limit) // precision
    n_limit = n * limit
    while True:
        x_old = x
        x = (x + n_limit // x) // 2
        if x == x_old:
            break
    return x


def Pi(limit):
    """
    limit -> pi
    Calculate PI up to the Nth digit using a Chudnovksy's series.
    """
    k = 1
    a_k = limit
    a_tot = limit
    b_tot = 0
    C = 640320
    C3_OVER_24 = C ** 3 // 24
    while True:
        a_k *= -(6 * k - 5) * (2 * k - 1) * (6 * k - 1)
        a_k //= k ** 3 * C3_OVER_24
        a_tot += a_k
        b_tot += k * a_k
        k += 1
        if a_k == 0:
            break
    total = 13591409 * a_tot + 545140134 * b_tot
    pi = (426880 * sqrt(10005 * limit, limit) * limit) // total
    return pi


def inputLimit():
    """
    <input> -> digit
    Get user's limit for PI
    Check for non negative number less than MAX_NUMBER (10000 by default)
    """
    while True:
        s = input("Enter the number of decimals to calculate to: ")
        try:
            digit = int(s)
            if digit >= MAX_NUMBER:
                print(f"Enter a number smaller than {MAX_NUMBER}")
            elif digit > 0:
                return digit
            else:
                print("Enter a positive integer.")
        except ValueError:
            print("Enter a positive integer.")


def main():
    """
    Main function
    - Get user's limit
    - Calculate result
    - Print result
    """

    print("-- Find Pi up to the Nth digit --\n")

    shouldContinue = True

    while shouldContinue:

        limit = inputLimit()
        pi = str(Pi(10 ** (limit * 10)))[:limit]
        pi = pi[0] + '.' + pi[1:]

        # show 40 digits by one line

        i = 0
        for digit in pi:
            print(digit, end='')
            i += 1
            if i == 40:
                print("")
                i = 0

        should = input("\nContinue (Y/[N]): ")
        if should.upper() != 'Y':
            shouldContinue = False
        else:
            shouldContinue = True


if __name__ == '__main__':
    main()

File: Numbers/Pi/Java/pi.java
import java.math.*;
import java.util.Scanner; import java.lang.*;

public class pi {

    public static BigDecimal pi =  BigDecimal.ZERO;
    public static BigDecimal denom1 =  BigDecimal.ONE;
    public static BigDecimal denom2 =  BigDecimal.ONE;
    public static BigDecimal term1 =  BigDecimal.ZERO;
    public static BigDecimal term2 =  BigDecimal.ZERO;

    public static void pi(int x)
    {
        // Nilakantha Series
        // pi = (3 + ((4/(i*(i+1)*(i+2))-(4/((i+2)*(i+3)*(i+4))))))...
        // calculate denominators, terms, then pi

        for(int i=2;i<1000;i+=4)
        {
            denom1 =  BigDecimal.ONE;
            denom1 = denom1.multiply(new BigDecimal(i));
            denom1 = denom1.multiply(new BigDecimal(i+1));
            denom1 = denom1.multiply(new BigDecimal(i+2));

            denom2 =  BigDecimal.ONE;
            denom2 = denom2.multiply(new BigDecimal(i+2));
            denom2 = denom2.multiply(new BigDecimal(i+3));
            denom2 = denom2.multiply(new BigDecimal(i+4));

            term1 = new BigDecimal("4").divide(denom1, 1000, RoundingMode.HALF_UP);
            term2 = new BigDecimal("-4").divide(denom2, 1000, RoundingMode.HALF_UP);

            pi = pi.add(term1);
            pi = pi.add(term2);
        }

        pi = pi.add(new BigDecimal (3));

        pi = pi.setScale(x,  BigDecimal.ROUND_HALF_UP);
    }

    public static void inputLimit()
    {
        // <input> -> digit
        System.out.println("Enter the number of decimals to calculate to: ");

        Scanner scanner = new Scanner(System.in);
        final int digit = Integer.parseInt(scanner.nextLine());

        pi(digit);
        System.out.println("Pi: " + pi);

        pi =  BigDecimal.ZERO;

        System.out.println("Enter '1' to continue: ");
        final int answer = Integer.parseInt(scanner.nextLine());

        if(answer == 1)
            inputLimit();
    }

    public static void main(String[] args)
    {
        // Main function
        System.out.println("-- Find Pi up to the Nth digit --\n");

        inputLimit();
    }
}